kvs-gbif/src/main/java/org/gbif/kvs/species/BackboneMatchByID.java

```diff
@@ -0,0 +1,165 @@
+
+import org.apache.commons.lang3.StringUtils;
+import org.gbif.api.vocabulary.OccurrenceIssue;
+import org.gbif.nameparser.NameParserGBIF;
+import org.gbif.nameparser.api.NameParser;
+import org.gbif.nameparser.api.ParsedName;
+import org.gbif.rest.client.species.ChecklistbankService;
+import org.gbif.rest.client.species.NameUsageSearchResponse;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Map;
+import java.util.Set;
+
+import static org.gbif.api.model.Constants.NUB_DATASET_KEY;
+import static org.gbif.api.vocabulary.OccurrenceIssue.*;
+
+/**
+ * A utility class to perform the logic of locating a backbone taxon from a variety of IDs.
+ * This will first make use of scientificNameID, then taxonID and then taxonConceptID, to locate a concept in the
+ * checklists configured for that ID prefix. If a backbone key is found, then it is returned, otherwise it moves
+ * on to the next identifier. While processing it will apply flags to indicate if the IDs are ignored, and will
+ * additionally flag if the scientificName appears different on the record to that in the checklist.
+ */
+public class BackboneMatchByID {
+  private static final Logger LOG = LoggerFactory.getLogger(BackboneMatchByID.class);
+
+  private static NameParser PARSER = new NameParserGBIF(2000, 0, 3);
+  private ChecklistbankService checklistbankService;
+  private Map<String, String> prefixMapping;
+  private Map<String, String> prefixToChecklistKey;
+
+  public BackboneMatchByID(ChecklistbankService checklistbankService, Map<String, String> prefixMapping, Map<String, String> prefixToChecklistKey) {
+    this.checklistbankService = checklistbankService;
+    this.prefixMapping = prefixMapping;
+    this.prefixToChecklistKey = prefixToChecklistKey;
+  }
+
+  Integer lookupBackboneKey(Identification identification, Set<OccurrenceIssue> issues) {
+    Integer backboneUsageKey;
+    String scientificName = identification.getScientificName();
+    String scientificNameID = identification.getScientificNameID();
+    String taxonID = identification.getTaxonID();
+    String taxonConceptID = identification.getTaxonConceptID();
+
+    LOG.debug("Attempting to locate a backbone taxon for scientificNameID[{}]", scientificNameID);
+    backboneUsageKey = lookupBackboneKey(scientificNameID, scientificName, issues, TAXON_MATCH_SCIENTIFIC_NAME_ID_IGNORED, SCIENTIFIC_NAME_ID_NOT_FOUND);
+
+    if (backboneUsageKey != null) {
+      LOG.debug("Backbone key[{}] found using scientificNameID[{}]. No further taxon IDs will be used", backboneUsageKey, scientificNameID);
+      issues.add(TAXON_MATCH_TAXON_ID_IGNORED);
+      issues.add(TAXON_MATCH_TAXON_CONCEPT_ID_IGNORED);
+    } else {
+      LOG.debug("Attempting to locate a backbone taxon for taxonID[{}]", taxonID);
+      backboneUsageKey = lookupBackboneKey(taxonID, scientificName, issues, TAXON_MATCH_TAXON_ID_IGNORED, TAXON_ID_NOT_FOUND);
+    }
+
+    if (backboneUsageKey != null) {
```

Comment:
Won't this still be true if it was found on line 48 with the scientificNameId? i.e. should this block be in the else block ending on line 57.